### PR TITLE
Resolves #2212 Add ApplyMigrationsAsync (and related async methods)

### LINKED ISSUE #2212: Migrations: Add ApplyMigrationsAsync (and related async methods)

### PR DESCRIPTION
**Pull request by laskoviymishka**
Add ApplyMigrationsAsync method to IMigrator interface and implement it. 
Add extension to call this method throw RelationalDatabaseFacadeExtensions.

### CODE REVIEW

File: src/EntityFramework.Relational/Migrations/Migrator.cs [hunk 1]
```diff
@@ -94,6 +95,11 @@ public virtual IReadOnlyList<Migration> GetUnappliedMigrations()
 
         public virtual bool HasPendingModelChanges() => _modelDiffer.HasDifferences(_migrationAssembly.LastModel, _model);
 
+        public async Task ApplyMigrationsAsync(string targetMigration = null)
+        {
+            await Task.Run(() => ApplyMigrations(targetMigration));
```
Comment (bricelam) [inline src/EntityFramework.Relational/Migrations/Migrator.cs]: This is a bad practice. (See [Async Best Practices](http://channel9.msdn.com/Events/TechEd/NorthAmerica/2014/DEV-B362).) We need a proper end-to-end async implementation.

Comment (laskoviymishka) [inline src/EntityFramework.Relational/Migrations/Migrator.cs]: Hi, re implemented this line. Also added async methods that uses in this apply migrations.

**Comment by laskoviymishka** [comment]
Add ExistsAsync and GetAppliedMigrationsAsync methods in history repository interface and implement it.
Reimplement ApplyMigrationsAsync in async way.

**Comment by rowanmiller** [comment]
Hey,
Having two code paths is the right thing to do here but at the moment there is a large amount of duplicated code that is not async. We would like to see this refactored so that the duplicated logic between the async/sync code paths is mostly just the parts that are actually async. What we are after is factoring out the code that does not change into helper methods etc. so that we don't have two copies of it.
~Rowan

**Comment by laskoviymishka** [comment]
Hi,
I Refactored migrator and history repositories. Extract shared code to helper methods.

Best regards,
Andrei Tserakhau

**Comment by laskoviymishka** [comment]
Fix merge conflicts.

**Comment by bricelam** [comment]
Hi, @laskoviymishka, I apologize for the delay in processing this. There has been a lot of churn in Migrations recently. Would you mind updating on top of the latest changes?

**Comment by laskoviymishka** [comment]
Hi, @bricelam sure. I update it and notify you when it will be completed. Sorry from my side also :sweat_smile:  was pretty busy.

**Comment by laskoviymishka** [comment]
Hi, @bricelam code was synced with latest changes. Could you please review it. I accidentally close PR  :), but reopen it again.

**Comment by bricelam** [comment]
Yikes: 301 changed files with 23,100 additions. I think something went wrong with the update...

**Comment by laskoviymishka** [comment]
Hm. Something goes wrong. Definitely. 
Maybe i should create new PR

**Comment by bricelam** [comment]
A new PR would be fine.

**Comment by laskoviymishka** [comment]
@bricelam Sorry again, my bad :cry: . Create new PR #3059 could you please review it? (it only 6 changed files with 253 additions and 73 deletions :+1: ).

**Pull request closed without merging**